File: 001-050/Problem004/prob4.go
package main

import (
	"os"
	"fmt"
	"strconv"
	"math"
)

// Given a number of digits, find the largest palindrome by multiplying all numbers
// in descending order
func approach1(num float64) {
	var max float64 = 0
	for i := math.Pow(10, num)-1; i > 99; i-- {
		for j := math.Pow(10, num)-1; j >= i; j-- {
			if isPalindrome(int(i*j)) && i*j > max {
				max = i*j
			}
		}
	}
	fmt.Println(max)
}

func reverse(str string) (result string) {
	for _, c := range str {
		result = string(c) + result
	}
	return
}

func isPalindrome(num int) bool {
	return strconv.Itoa(num) == reverse(strconv.Itoa(num))
}

func main() {
	num, err := strconv.Atoi(os.Args[1])
	if err != nil {
		fmt.Println(err)
	}

	approach1(float64(num))
}

File: 001-050/Problem004/prob4.py
"""
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
"""

from MathLib.digits import isPalindrome

def main():
    max = 0
    for m in range((10**3)-1, 100-1, -1):
        for n in range((10**3)-1, m-1, -1):
            if isPalindrome(m*n) and m*n > max:
                max = m*n
    print(max)

if __name__ == "__main__":
    main()